### DISCUSSION TITLE
when i call await request.body() the request hangs up

### DISCUSSION THREAD

**Discussion by Teklvtn** [fastapi/fastapi#9678] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request
from pydantic import BaseModel
import uvicorn


class Test(BaseModel):
    s: int


app = FastAPI()


@app.middleware("http")
async def log_request_body(request: Request, call_next):
    # body = b""
    # async for chunk in request.stream():
    #     body += chunk
    # print(f"Request body: {body.decode()}")
    body_content = await request.body()
    print(f"Request body: {body_content.decode()}")
    response = await call_next(request)
    return response


@app.post('/')
def root():
    return 1


@app.post('/ad')
def test(test: Test):
    return 1

if __name__ == '__main__':
    uvicorn.run(app)
```

### Description

I use middleware in the project to process the request object. Due to unknown problems in the project, I need to see whether it is a problem with the request body or an internal processing flow problem, so I added a middleware and passed await request.body( ) to read

But I don’t know what caused the execution of the middleware to stop when response = await call_next(request). According to the log printing situation, it should not enter the specific endpoint processing flow

However, some endpoints are executed normally, and for those endpoints that actually have a body, the request hangs up.

In the example code, in the provided swagger page, such a situation will occur when requesting http://127.0.0.1:8000/ad, and the page will always be in the loading state. This may be a bug, and I have no way to confirm it. Is it my problem, or a bug

### Operating System

Windows

### Operating System Details

win10

### FastAPI Version

0.95.1

### Python Version

3.8.9

### Additional Context

Maybe it's a bug in routing handling?

**Comment by Teklvtn** [comment]
In the end I solved it with global dependencies, but I'm still curious why exactly this is happening in middleware

**Comment by Kludex** [comment]
Can you try to use pure Starlette (the latest version), and see if it's solved? FastAPI needs to bump Starlette.

**Reply by Teklvtn** [reply]
let me try

**Reply by Teklvtn** [reply]
I specifically tried the highest version, but it seems that some additional parameters are added, and there is no such parameter (debug) in FastAPI, so I guess it may be because the version of FastAPI has not kept up, so I installed the latest version for this, or this phenomenon

**Reply by Teklvtn** [reply]
Do you have any ideas on how to deal with it? I am curious why this happens, it is feasible for me to read the body through dependencies @Kludex

**Reply by Teklvtn** [reply]
After debugging, I found that the message = await recv_stream.receive() in line 78 of venv/Lib/site-packages/starlette/middleware/base.py is waiting here, so the control of the coroutine is submitted, resulting in The follow-up call_next is broken


The debug method is to put a breakpoint on the line await call_next(request), and then step into the function

**Reply by Teklvtn** [reply]
```
@dataclass(eq=False)
class MemoryObjectReceiveStream(Generic[T_co], ObjectReceiveStream[T_co]):
    _state: MemoryObjectStreamState[T_co]
    _closed: bool = field(init=False, default=False)

    def __post_init__(self) -> None:
        self._state.open_receive_channels += 1

    def receive_nowait(self) -> T_co:
        """
        Receive the next item if it can be done without waiting.

        :return: the received item
        :raises ~anyio.ClosedResourceError: if this send stream has been closed
        :raises ~anyio.EndOfStream: if the buffer is empty and this stream has been
            closed from the sending end
        :raises ~anyio.WouldBlock: if there are no items in the buffer and no tasks
            waiting to send

        """
        if self._closed:
            raise ClosedResourceError

        if self._state.waiting_senders:
            # Get the item from the next sender
            send_event, item = self._state.waiting_senders.popitem(last=False)
            self._state.buffer.append(item)
            send_event.set()

        if self._state.buffer:
            return self._state.buffer.popleft()
        elif not self._state.open_send_channels:
            raise EndOfStream

        raise WouldBlock
```
Eventually, the coroutine will die due to advancing to raise WouldBlock @tiangolo 

It seems that after the body of the request is read once, the coroutine will die later because it cannot be read. By relying on it, it may be because the body stream in the middleware is read independently, so it doesn't matter. Here is Don't you need to recreate a stream object?

**Comment by mxzgn** [comment]
Me too, how to solve

**Reply by Teklvtn** [reply]
Since I only need to do some pre-operations before entering the routing processing function, I solved this problem with dependencies. It is feasible to obtain the body() byte stream through await in the dependencies, and it will not affect the execution of call_next

**Comment by Teklvtn** [comment]
```
from starlette.applications import Starlette
from starlette.responses import PlainTextResponse
from starlette.routing import Route, WebSocketRoute
import uvicorn
from starlette.middleware import Middleware
from starlette.middleware.base import BaseHTTPMiddleware


def homepage(request):
    return PlainTextResponse('Hello, world!')


def startup():
    print('Ready to go')


routes = [
    Route('/', homepage, methods=['POST']),
]


class CustomHeaderMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        print(await request.body())
        response = await call_next(request)
        response.headers['Custom'] = 'Example'
        return response


app = Starlette(debug=True, routes=routes, on_startup=[startup], middleware=[Middleware(CustomHeaderMiddleware)])

if __name__ == '__main__':
    uvicorn.run(app)
```
The above is my test code, and the following is an example of my request to the demo, and the result is a normal response
```
import requests

url = "http://127.0.0.1:8000"

payload={}
files=[

]
headers = {}

response = requests.request("POST", url, headers=headers, data=payload, files=files)

print(response.text)
```
